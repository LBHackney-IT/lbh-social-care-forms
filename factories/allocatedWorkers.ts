import { Factory } from 'fishery';

import { Allocation } from 'types';

export const allocationFactory = Factory.define<Allocation>(({ sequence }) => ({
  id: sequence,
  allocatedWorkerTeam: 'Safeguarding and Reviewing Team',
  allocatedWorker: 'Officer Name',
  allocationEndDate: '2019-03-28 00:00:00',
  allocationStartDate: '2019-03-28 00:00:00',
  workerType: 'Consultant Social Worker',
  personId: 1234,
  caseStatus: 'Open',
  personName: 'foo',
  personAddress: 'the address',
  personDateOfBirth: '2020-03-20',
  ragRating: 'amber',
}));

export const mockedAllocation = allocationFactory.build();
export const mockedAllocations = [allocationFactory.build()];
