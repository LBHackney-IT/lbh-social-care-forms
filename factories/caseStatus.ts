import { Factory } from 'fishery';
import {
  CaseStatus,
  CaseStatusFields,
  AddCaseStatusFormData,
  CaseStatusAnswerDisplay,
} from 'types';

export const mockedCaseStatusFactory = Factory.define<CaseStatus>(
  ({ sequence }) => ({
    id: sequence,
    type: 'CIN',
    answers: [mockedStatusField.build()],
    startDate: '2021-01-01T02:00:00Z',
    endDate: '2021-12-01T02:00:00Z',
    notes: 'a lot to note',
  })
);

export const mockedCaseStatusAddRequest = Factory.define<AddCaseStatusFormData>(
  ({ sequence }) => ({
    personId: sequence,
    type: 'CIN',
    fields: [{ name: 'test', selected: 'test' }],
    startDate: '2021-01-01T02:00:00Z',
    endDate: '2021-12-01T02:00:00Z',
    notes: 'a lot to note',
    createdby: 'test@testington.com',
  })
);

export const mockedStatusField = Factory.define<CaseStatusFields>(() => ({
  option: 'placementType',
  value: 'K1',
  startDate: '2020-08-01',
  createdAt: '2021-10-11T10:54:32Z',
}));

export const mockedCaseStatusAnswers = Factory.define<CaseStatusAnswerDisplay>(
  () => ({
    startDate: '2021-09-10',
    endDate: '2021-10-10',
    status: [mockedStatusField.build()],
  })
);
