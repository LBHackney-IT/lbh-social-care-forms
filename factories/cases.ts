import { Factory } from 'fishery';

import { Case } from 'types';

export const caseFactory = Factory.define<Case>(({ sequence }) => ({
  recordId: sequence.toString(),
  personId: 123,
  dateOfEvent: '25/10/2020 13:49:43',
  formName: 'foorm',
  officerEmail: 'Fname.Lname@hackney.gov.uk',
  caseFormData: {
    note: 'I am the note',
    first_name: 'Foo',
    last_name: 'Bar',
    timestamp: '11/03/2021 16:51:56',
    mosaic_id: 123,
    form_name_overall: 'foo',
    worker_email: 'worker@hackney.gov.uk',
    form_name: 'Foo bar',
    context_flag: 'A',
    date_of_event: '25/10/2020 13:49:43',
  },
  deleted: false,
  isImported: false,
}));

export const mockedNote = caseFactory.build({ caseFormUrl: 'https://foo.bar' });
export const mockedAllocationNote = caseFactory.build({
  formName: 'Worker allocated',
  caseFormData: {
    allocation_id: 321,
    created_by: 'foo@hackney.gov.uk',
    form_name_overall: 'API_Allocation',
    form_name: 'Worker allocated',
  },
});
export const mockedDeallocationNote = caseFactory.build({
  formName: 'Worker allocated',
  caseFormData: {
    allocation_id: 321,
    created_by: 'foo@hackney.gov.uk',
    deallocation_reason: 'deallocated because',
    form_name_overall: 'API_Deallocation',
    form_name: 'Worker deallocated',
  },
});
export const mockedCaseNote = caseFactory.build({
  caseFormData: {
    form_name_overall: 'ASC_case_note',
    case_note_title: 'i am a case title',
  },
});
export const mockedDeletedCaseNote = caseFactory.build({
  caseFormData: {
    form_name_overall: 'ASC_case_note',
    case_note_title: 'i am a case title',
  },
  formName: 'adult-case-note',
  deleted: true,
});
export const mockedWarningNoteCase = caseFactory.build({
  formName: 'Warning Note',
  caseFormData: {
    form_name_overall: 'API_WarningNote',
    warning_note_id: 456,
  },
});
