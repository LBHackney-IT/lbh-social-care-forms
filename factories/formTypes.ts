import { Factory } from 'fishery';
import { FormValue, FormStatusValue } from 'types';

export const mockedFormValueFactory = Factory.define<FormValue>(() => ({
  name: 'test',
  description: 'description',
  options: [mockedFormStatusValueFactory.build()],
}));

export const mockedFormStatusValueFactory = Factory.define<FormStatusValue>(
  () => ({
    name: 'N1',
    description: 'Noo 1',
  })
);

export const mockedFormValue = [mockedFormValueFactory.build()];
