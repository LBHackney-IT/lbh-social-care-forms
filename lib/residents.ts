import axios from 'axios';

import type { LegacyResident, Resident, ResidentsAPI, User } from 'types';
import { getAuditingParams } from '../utils/auditing';

const ENDPOINT_API = process.env.ENDPOINT_API;
const AWS_KEY = process.env.AWS_KEY;

const headers = { 'x-api-key': AWS_KEY };

interface ResidentBE extends LegacyResident {
  addressList: Array<{
    displayAddressFlag: 'Y' | 'N';
    addressLine1: string;
    postCode: string;
    uprn: string;
  }>;
}

const sanitiseResidentData = (residents: ResidentBE[]): LegacyResident[] =>
  residents?.map(({ addressList, ...resident }: ResidentBE): LegacyResident => {
    const address = addressList?.find(
      ({ displayAddressFlag }) => displayAddressFlag === 'Y'
    );
    return {
      ...resident,
      address: address && {
        address: address.addressLine1,
        postcode: address.postCode,
        uprn: address.uprn,
      },
    };
  });

export const getResidents = async (
  params: Record<string, unknown>
): Promise<ResidentsAPI> => {
  const { data } = await axios.get(`${ENDPOINT_API}/residents`, {
    headers,
    params,
  });
  return { ...data, residents: sanitiseResidentData(data.residents) };
};

export const getResident = async (
  personId: number,
  user: User
): Promise<Resident> => {
  const { data } = await axios.get(`${ENDPOINT_API}/residents/${personId}`, {
    headers,
    params: getAuditingParams(user),
  });
  return data;
};

export const normalisePhoneInput = (
  formData: Resident
): Record<string, unknown> => ({
  ...formData,
  phoneNumbers: formData.phoneNumbers
    ? formData.phoneNumbers.map((phone) => ({
        ...phone,
        type: phone.type === '' ? 'main' : phone.type,
      }))
    : null,
});

export const addResident = async (
  formData: Resident
): Promise<ResidentsAPI> => {
  const { data } = await axios.post(
    `${ENDPOINT_API}/residents`,
    normalisePhoneInput(formData),
    {
      headers: { ...headers, 'Content-Type': 'application/json' },
    }
  );
  return data;
};

export const updateResident = async (formData: Resident): Promise<Resident> => {
  await axios.patch(
    `${ENDPOINT_API}/residents`,
    normalisePhoneInput(formData),
    {
      headers: { ...headers, 'Content-Type': 'application/json' },
    }
  );
  return formData;
};
