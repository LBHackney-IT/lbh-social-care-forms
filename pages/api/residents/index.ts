import { StatusCodes } from 'http-status-codes';

import { getResidents, addResident } from 'lib/residents';
import { isAuthorised } from 'utils/auth';
import { apiHandler } from 'lib/apiHandler';
import { middleware as csrfMiddleware } from 'lib/csrfToken';

import type { NextApiRequest, NextApiResponse, NextApiHandler } from 'next';
import { AxiosError } from 'axios';

const endpoint: NextApiHandler = async (
  req: NextApiRequest,
  res: NextApiResponse
) => {
  const user = isAuthorised(req);
  if (!user) {
    res.status(StatusCodes.UNAUTHORIZED);
    return;
  }
  if (!user.isAuthorised) {
    res.status(StatusCodes.FORBIDDEN);
    return;
  }
  switch (req.method) {
    case 'GET':
      try {
        const data = await getResidents(req.query);
        res.status(StatusCodes.OK).json(data);
      } catch (error) {
        console.error(
          'Residents get error:',
          (error as AxiosError)?.response?.data
        );
        (error as AxiosError)?.response?.status === StatusCodes.NOT_FOUND
          ? res
              .status(StatusCodes.NOT_FOUND)
              .json({ message: 'Residents Not Found' })
          : res
              .status(StatusCodes.INTERNAL_SERVER_ERROR)
              .json({ message: 'Unable to get the Residents' });
      }
      break;

    case 'POST':
      try {
        const data = await addResident(req.body);
        res.status(StatusCodes.OK).json(data);
      } catch (error) {
        console.error(
          'Resident post error:',
          (error as AxiosError)?.response?.data
        );
        console.error('Resident post request:', req);
        res
          .status(StatusCodes.INTERNAL_SERVER_ERROR)
          .json({ message: 'Unable to add resident' });
      }
      break;

    default:
      res
        .status(StatusCodes.BAD_REQUEST)
        .json({ message: 'Invalid request method' });
  }
};

export default apiHandler(csrfMiddleware(endpoint));
