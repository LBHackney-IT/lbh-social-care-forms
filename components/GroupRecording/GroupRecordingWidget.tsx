import { Resident, LegacyResident } from './../../types';
import { useState, useEffect } from 'react';
import PersonWidget from './../PersonWidget/PersonWidget';
import Dialog from './../Dialog/Dialog';
import s from './GroupRecordingWidget.module.scss';
import PersonSelect from './../PersonSelect/PersonSelect';
import axios from 'axios';
import { tokenFromMeta } from 'lib/csrfToken';
import { useResidents as useIDSearch } from 'utils/api/residents';
import { useResidents as useFirstNameSearch } from 'utils/api/residents';
import { useResidents as useLastNameSearch } from 'utils/api/residents';

interface Props {
  initialPeople: Resident[];
  submissionId: string;
}

const safeId = (person: { id?: number; mosaicId?: string }): number =>
  person?.id ? person.id : parseInt(String(person.mosaicId));

const GroupRecordingWidget = ({
  initialPeople,
  submissionId,
}: Props): React.ReactElement => {
  const [people, setPeople] =
    useState<(Resident | LegacyResident)[]>(initialPeople);
  const [open, setOpen] = useState<number | false>(0);
  const [dialogOpen, setDialogOpen] = useState<boolean>(false);
  const [idToAdd, setIdToAdd] = useState<number>(-1);

  const [searchQuery, setSearchQuery] = useState('');

  const { data: idResults } = useIDSearch({
    mosaic_id: searchQuery,
  });
  const { data: firstNameResults } = useFirstNameSearch({
    first_name: searchQuery,
  });
  const { data: lastNameResults } = useLastNameSearch({
    last_name: searchQuery,
  });

  const results: (Resident | LegacyResident)[] = [
    ...((lastNameResults?.[0]?.residents || []) as (
      | Resident
      | LegacyResident
    )[]),
    ...((idResults?.[0]?.residents || []) as (Resident | LegacyResident)[]),
    ...((firstNameResults?.[0]?.residents || []) as (
      | Resident
      | LegacyResident
    )[]),
  ]
    // remove duplicates
    .reduce((unique, p1) => {
      if (!unique.some((p2) => safeId(p1) === safeId(p2))) unique.push(p1);
      return unique;
    }, [] as (Resident | LegacyResident)[])
    // don't show anyone already on the list
    .filter((person) => {
      if ('id' in person) {
        return !people
          .map((existingPerson) => {
            if ('id' in existingPerson) {
              return existingPerson.id;
            } else {
              return parseInt(String(existingPerson.mosaicId));
            }
          })
          .includes(person.id);
      } else {
        return !people
          .map((existingPerson) => {
            if ('id' in existingPerson) {
              return existingPerson.id;
            } else {
              return parseInt(String(existingPerson.mosaicId));
            }
          })
          .includes(parseInt(String(person.mosaicId)));
      }
    })
    // only show nine max
    .slice(0, 9);

  function mapPeopleToIds(person: LegacyResident | Resident) {
    if ('id' in person) {
      return person.id;
    } else {
      return parseInt(String(person.mosaicId));
    }
  }

  useEffect(() => {
    const residents = people.map((person) => mapPeopleToIds(person));

    axios.patch(
      `/api/submissions/${submissionId}`,
      {
        residents,
      },
      {
        headers: {
          'XSRF-TOKEN': tokenFromMeta(),
        },
      }
    );
  }, [people, submissionId]);

  const handleAdd = (): void => {
    if (idToAdd) {
      const result =
        results &&
        results.find((resident) => {
          if ('id' in resident) {
            return resident.id === idToAdd;
          } else {
            return resident.mosaicId === String(idToAdd);
          }
        });

      if (result !== undefined) {
        setPeople([...people, result]);
      } else {
        setPeople([...people]);
      }
      setOpen(people.length);
      setDialogOpen(false);
      setSearchQuery('');
    }
  };

  const handleRemove = (idToRemove: number): void => {
    function filterPeople(person: Resident | LegacyResident) {
      if ('id' in person) {
        return person.id !== idToRemove;
      } else {
        return person.mosaicId !== String(idToRemove);
      }
    }

    const newPeople: (Resident | LegacyResident)[] = people.filter((person) =>
      filterPeople(person)
    );

    setPeople(newPeople);
    setOpen(newPeople.length - 1);
  };

  return (
    <section>
      <h3 className="govuk-visually-hidden">People</h3>
      {people.map((person, i) => (
        <PersonWidget
          person={person}
          key={i}
          index={i}
          grouped={people.length > 1}
          onRemove={handleRemove}
          open={open === i}
          setOpen={setOpen}
        />
      ))}
      <button
        className={`lbh-link ${s.addPersonLink}`}
        onClick={() => setDialogOpen(true)}
      >
        <svg
          width="15"
          height="15"
          viewBox="0 0 15 15"
          fill="none"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect x="7" width="1" height="15" fill="#025EA6" />
          <rect
            x="15"
            y="7"
            width="1"
            height="15"
            transform="rotate(90 15 7)"
            fill="#025EA6"
          />
        </svg>
        Link another person
      </button>

      <Dialog
        title="Link a person"
        isOpen={dialogOpen}
        onDismiss={() => setDialogOpen(false)}
      >
        <div className="govuk-form-group lbh-form-group">
          <label className="govuk-label lbh-label" htmlFor="query">
            Search by name or social care ID
          </label>
          <div className="lbh-search-box" style={{ marginTop: 0 }}>
            <input
              className="govuk-input lbh-input"
              value={searchQuery}
              onChange={(e) => setSearchQuery(e.target.value)}
              id="query"
              name="query"
              type="search"
              placeholder="eg. 123456"
            />
            <button className="lbh-search-box__action">
              <span className="govuk-visually-hidden">Search</span>
              <svg width="16" height="16" viewBox="0 0 16 16" fill="none">
                <path
                  fillRule="evenodd"
                  clipRule="evenodd"
                  d="M9.6999 10.6C12.0747 10.6 13.9999 8.67482 13.9999 6.3C13.9999 3.92518 12.0747 2 9.6999 2C7.32508 2 5.3999 3.92518 5.3999 6.3C5.3999 8.67482 7.32508 10.6 9.6999 10.6ZM9.6999 12.6C13.1793 12.6 15.9999 9.77939 15.9999 6.3C15.9999 2.82061 13.1793 0 9.6999 0C6.22051 0 3.3999 2.82061 3.3999 6.3C3.3999 9.77939 6.22051 12.6 9.6999 12.6Z"
                  fill="#0B0C0C"
                />
                <path
                  fillRule="evenodd"
                  clipRule="evenodd"
                  d="M6.70706 10.7071L1.70706 15.7071L0.292847 14.2929L5.29285 9.29289L6.70706 10.7071Z"
                  fill="#0B0C0C"
                />
              </svg>
            </button>
          </div>
        </div>

        {searchQuery && results ? (
          <PersonSelect
            people={results}
            label="Matching people"
            idToAdd={idToAdd}
            setIdToAdd={setIdToAdd}
          />
        ) : (
          <></>
        )}

        <button
          className="govuk-button lbh-button"
          onClick={handleAdd}
          disabled={idToAdd === -1}
        >
          Add person
        </button>
      </Dialog>
    </section>
  );
};

export default GroupRecordingWidget;
