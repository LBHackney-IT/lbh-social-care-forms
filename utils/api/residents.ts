import axios from 'axios';
import useSWR, { useSWRInfinite, SWRResponse, SWRInfiniteResponse } from 'swr';

import { getInfiniteKey } from 'utils/api';

import type { Resident, ErrorAPI } from 'types';

export const useResidents = (
  params: Record<string, unknown>,
  invoke = true
): SWRInfiniteResponse<Record<string, unknown>, Error> =>
  useSWRInfinite(
    // @ts-ignore
    invoke ? getInfiniteKey('/api/residents', 'residents', params) : null
  );

export const useResident = (
  personId: number
): SWRResponse<Resident, ErrorAPI> => useSWR(`/api/residents/${personId}`);

export const addResident = async (
  formData: Record<string, unknown>
): Promise<Record<string, unknown>> => {
  const { data } = await axios.post(`/api/residents`, formData);
  return { ref: data?.id, data };
};

export const updateResident = async (
  personId: number,
  formData: Record<string, unknown>
): Promise<Record<string, unknown>> => {
  const { data } = await axios.patch(`/api/residents/${personId}`, formData);
  return { ref: data?.id, data };
};

export const patchResident = async (formData: {
  id: number;
  createdBy: string;
  [key: string]: unknown;
}): Promise<Record<string, unknown>> => {
  const { data } = await axios.patch(`/api/residents`, formData);
  return { ref: data?.id, data };
};
