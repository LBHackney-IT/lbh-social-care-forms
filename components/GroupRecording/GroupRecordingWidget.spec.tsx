import GroupRecordingWidget from './GroupRecordingWidget';
import { fireEvent, render, screen } from '@testing-library/react';
import { mockedResident, residentFactory } from 'factories/residents';
import { useResidents } from 'utils/api/residents';
import * as CSRFToken from 'lib/csrfToken';
import axios from 'axios';

jest.mock('next/router', () => ({
  useRouter: () => ({
    query: { id: '1' },
    asPath: 'path',
  }),
}));

jest.mock('utils/api/residents', () => ({
  useResidents: jest.fn(),
}));
jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

describe('GroupRecordingWidget', () => {
  const mockedResident1 = residentFactory.build({
    id: 1,
    firstName: 'Foo',
    lastName: 'Bar',
  });

  const mockedResident2 = residentFactory.build({
    id: 2,
    firstName: 'Bar',
    lastName: 'Baz',
  });

  beforeEach(() => {
    (useResidents as jest.Mock).mockReturnValue({
      size: 2,
      data: [
        {
          residents: [mockedResident1, mockedResident2],
        },
      ],
    });
  });
  const mock = jest.spyOn(CSRFToken, 'tokenFromMeta'); // spy on otherFn
  mock.mockReturnValue('TEST_XSRF');

  it('should render properly', () => {
    const { asFragment } = render(
      <GroupRecordingWidget submissionId="1" initialPeople={[mockedResident]} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('shows one person initially', () => {
    render(
      <GroupRecordingWidget
        submissionId="1"
        initialPeople={[mockedResident1]}
      />
    );

    expect(
      screen.getByText(
        `${mockedResident1.firstName} ${mockedResident1.lastName}`
      )
    );
  });

  it('can show multiple people', () => {
    render(
      <GroupRecordingWidget
        submissionId="1"
        initialPeople={[mockedResident1, mockedResident2]}
      />
    );

    expect(
      screen.getByText(
        `${mockedResident1.firstName} ${mockedResident1.lastName}`
      )
    );
    expect(
      screen.getByText(
        `${mockedResident2.firstName} ${mockedResident2.lastName}`
      )
    );
  });

  it('shows an option to remove a person if multiple people are assigned to the group', () => {
    render(
      <GroupRecordingWidget
        submissionId="1"
        initialPeople={[mockedResident1, mockedResident2]}
      />
    );

    expect(screen.getAllByText('Remove').length).toBe(2);
  });

  it('does not show an option to remove a person if a single person is assigned to the group', () => {
    render(
      <GroupRecordingWidget
        submissionId="1"
        initialPeople={[mockedResident1]}
      />
    );

    expect(screen.queryByText('Remove')).toBeFalsy();
  });

  it('does not show a resident can be added if they already belong to the group', () => {
    render(
      <GroupRecordingWidget
        submissionId="1"
        initialPeople={[mockedResident1]}
      />
    );

    fireEvent.click(screen.getByText('Link another person'));

    fireEvent.change(screen.getByPlaceholderText('eg. 123456'), {
      target: { value: mockedResident1.id },
    });
    const personToAdd = screen.queryByLabelText(
      `${mockedResident1.firstName} ${mockedResident1.lastName}`
    ) as HTMLInputElement;
    expect(personToAdd).toBeNull();
  });

  it('shows a resident can be added if they do not belong to the group', () => {
    render(
      <GroupRecordingWidget
        submissionId="1"
        initialPeople={[mockedResident1]}
      />
    );

    fireEvent.click(screen.getByText('Link another person'));

    fireEvent.change(screen.getByPlaceholderText('eg. 123456'), {
      target: { value: mockedResident2.id },
    });
    const personToAdd = screen.getByText(
      `${mockedResident2.firstName} ${mockedResident2.lastName}`
    ) as HTMLInputElement;
    expect(personToAdd).toBeTruthy();
  });

  it('sets checked to true after selecting a resident to add', () => {
    render(
      <GroupRecordingWidget
        submissionId="1"
        initialPeople={[mockedResident1]}
      />
    );

    fireEvent.click(screen.getByText('Link another person'));

    fireEvent.change(screen.getByPlaceholderText('eg. 123456'), {
      target: { value: mockedResident2.id },
    });

    const residentToSelect = screen.getByLabelText(
      `${mockedResident2.firstName} ${mockedResident2.lastName}`
    ) as HTMLInputElement;
    expect(residentToSelect.checked).toBeFalsy();
    fireEvent.click(residentToSelect);
    expect(residentToSelect.checked).toBeTruthy();
  });

  it('add person button disabled if no resident is selected', () => {
    render(
      <GroupRecordingWidget
        submissionId="1"
        initialPeople={[mockedResident1]}
      />
    );

    fireEvent.click(screen.getByText('Link another person'));

    const addPersonButton = screen.getByText('Add person') as HTMLButtonElement;
    expect(addPersonButton.disabled).toBeTruthy();
  });

  it('add person button enabled if a resident is selected', () => {
    render(
      <GroupRecordingWidget
        submissionId="1"
        initialPeople={[mockedResident1]}
      />
    );

    fireEvent.click(screen.getByText('Link another person'));
    fireEvent.change(screen.getByPlaceholderText('eg. 123456'), {
      target: { value: mockedResident2.id },
    });
    fireEvent.click(
      screen.getByLabelText(
        `${mockedResident2.firstName} ${mockedResident2.lastName}`
      )
    );

    const addPersonButton = screen.getByText('Add person') as HTMLButtonElement;
    expect(addPersonButton.disabled).toBeFalsy();
  });

  it('makes a PATCH request with our new resident IDs after adding a resident', () => {
    render(
      <GroupRecordingWidget
        submissionId="1"
        initialPeople={[mockedResident1]}
      />
    );

    fireEvent.click(screen.getByText('Link another person'));

    fireEvent.change(screen.getByPlaceholderText('eg. 123456'), {
      target: { value: mockedResident2.id },
    });
    fireEvent.click(
      screen.getByLabelText(
        `${mockedResident2.firstName} ${mockedResident2.lastName}`
      )
    );

    fireEvent.click(screen.getByText('Add person'));

    expect(mockedAxios.patch).toHaveBeenLastCalledWith(
      '/api/submissions/1',
      {
        residents: [mockedResident1.id, mockedResident2.id],
      },
      { headers: { 'XSRF-TOKEN': 'TEST_XSRF' } }
    );
  });

  it('displays the newly added resident', () => {
    render(
      <GroupRecordingWidget
        submissionId="1"
        initialPeople={[mockedResident1]}
      />
    );

    fireEvent.click(screen.getByText('Link another person'));

    fireEvent.change(screen.getByPlaceholderText('eg. 123456'), {
      target: { value: mockedResident2.id },
    });
    fireEvent.click(
      screen.getByLabelText(
        `${mockedResident2.firstName} ${mockedResident2.lastName}`
      )
    );

    fireEvent.click(screen.getByText('Add person'));

    expect(
      screen.getByText(
        `${mockedResident2.firstName} ${mockedResident2.lastName}`
      )
    ).toBeTruthy();
  });

  it('makes a PATCH request with our new resident IDs after removing a resident', () => {
    render(
      <GroupRecordingWidget
        submissionId="1"
        initialPeople={[mockedResident1, mockedResident2]}
      />
    );

    fireEvent.click(screen.getAllByText('Remove')[0]);

    expect(mockedAxios.patch).toHaveBeenLastCalledWith(
      '/api/submissions/1',
      {
        residents: [mockedResident2.id],
      },
      { headers: { 'XSRF-TOKEN': 'TEST_XSRF' } }
    );
  });

  it('removes the removed resident from the group recording display', () => {
    render(
      <GroupRecordingWidget
        submissionId="1"
        initialPeople={[mockedResident1, mockedResident2]}
      />
    );

    fireEvent.click(screen.getAllByText('Remove')[0]);

    expect(
      screen.queryByText(
        `${mockedResident1.firstName} ${mockedResident1.lastName}`
      )
    ).toBeNull();
  });
});
