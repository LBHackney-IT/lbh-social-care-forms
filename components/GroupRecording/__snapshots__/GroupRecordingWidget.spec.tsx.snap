// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GroupRecordingWidget should render properly 1`] = `
<DocumentFragment>
  <section>
    <h3
      class="govuk-visually-hidden"
    >
      People
    </h3>
    <aside
      class="aside"
    >
      <h2
        class="lbh-heading-h3 title"
      >
        Foo Bar
      </h2>
      <p
        class="lbh-body-s paragraph"
      >
        Born 13 Nov 2020
      </p>
      <p
        class="lbh-body-s paragraph"
        data-testid="resident-display-address"
      >
        sjakdjlk
        <br />
        hdsadjk
      </p>
    </aside>
    <button
      class="lbh-link addPersonLink"
    >
      <svg
        fill="none"
        height="15"
        viewBox="0 0 15 15"
        width="15"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="#025EA6"
          height="15"
          width="1"
          x="7"
        />
        <rect
          fill="#025EA6"
          height="15"
          transform="rotate(90 15 7)"
          width="1"
          x="15"
          y="7"
        />
      </svg>
      Link another person
    </button>
  </section>
</DocumentFragment>
`;
